packages/@react-stately/layout/src/TableLayout.ts

```diff
@@ -324,35 +340,27 @@ export class TableLayout<T> extends ListLayout<T> {
         let firstVisibleCell = this.binarySearch(node.children, rect.topLeft, 'x');
         let lastVisibleCell = this.binarySearch(node.children, rect.topRight, 'x');
         let stickyIndex = 0;
-        // Check to see if a persisted key exists before the visible cells
-        // This is for keeping focus on a cell that scrolls out of view
-        for (let h = 0; h < firstVisibleCell; h++) {
-          if (this.virtualizer.isPersistedKey(node.children[h])) {
-            res.push(node.children[h].layoutInfo);
-          }
+
+        // Add persisted/sticky cells before the visible cells.
+        let persistedCellIndices = this.persistedIndices.get(node.layoutInfo.key) || this.stickyColumnIndices;
+        while (stickyIndex < persistedCellIndices.length && persistedCellIndices[stickyIndex] < firstVisibleCell) {
+          let idx = persistedCellIndices[stickyIndex];
+          res.push(node.children[idx].layoutInfo);
+          stickyIndex++;
         }
+
         for (let i = firstVisibleCell; i <= lastVisibleCell; i++) {
-          // Sticky columns and row headers are always in the DOM. Interleave these
-          // with the visible range so that they are in the right order.
-          if (stickyIndex < this.stickyColumnIndices.length) {
-            let idx = this.stickyColumnIndices[stickyIndex];
-            while (idx < i) {
-              res.push(node.children[idx].layoutInfo);
-              idx = this.stickyColumnIndices[stickyIndex++];
-            }
+          // Skip sticky cells that overlap with visible cells.
+          while (stickyIndex < persistedCellIndices.length && persistedCellIndices[stickyIndex] < i) {
+            stickyIndex++;
           }
 
           res.push(node.children[i].layoutInfo);
         }
-        // Check to see if a persisted key exists after the visible cells
-        for (let j = lastVisibleCell; j < node.children.length; j++) {
-          if (this.virtualizer.isPersistedKey(node.children[j])) {
-            res.push(node.children[j].layoutInfo);
-          }
-        }
 
-        while (stickyIndex < this.stickyColumnIndices.length) {
-          let idx = this.stickyColumnIndices[stickyIndex++];
+        // Add any remaining sticky cells after the visible cells.
+        while (stickyIndex < persistedCellIndices.length) {
+          let idx = persistedCellIndices[stickyIndex++];
```

Comment:
given that the last thing you did in the previous while loop was increment stickyIndex, might this skip an item unintentionally?

Reply:
this comment isn't showing up in the actual code diff which is strange, but I think it's ok because the loop checks whether `persistedCellIndices[stickyIndex] < i` before incrementing.